### PR TITLE
Add process env NEXT_RUNTIME

### PR DESCRIPTION
**Pull request by huozhi**
* To help determine the current running environment is in server nodejs / edge runtime
* Remove usage of `process.browser`

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 138d4e233701a73581f81e8fb6726f8b5102e412

`yarn testheadless test/production/middleware-environment-variables-in-node-server-reflect-the-usage-inference/index.test.ts` 
- middleware environment variables in node server reflect the usage inference > limits process.env to only contain env vars that are inferred from usage

<details>
<summary>Expand output</summary>

  ● middleware environment variables in node server reflect the usage inference › limits process.env to only contain env vars that are inferred from usage

    expect(received).toEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 1

      Object {
        "canBeInferred": "can-be-inferred",
        "rest": Object {
          "CAN_BE_INFERRED": "can-be-inferred",
          "X_CUSTOM_HEADER": "x-custom-header",
    +     "__NEXT_RUNTIME": "edge",
        },
      }

      40 |       parsed = JSON.parse(html)
      41 |     }).not.toThrow()
    > 42 |     expect(parsed).toEqual({
         |                    ^
      43 |       canBeInferred: 'can-be-inferred',
      44 |       rest: {
      45 |         CAN_BE_INFERRED: 'can-be-inferred',

      at Object.<anonymous> (production/middleware-environment-variables-in-node-server-reflect-the-usage-inference/index.test.ts:42:20)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 138d4e233701a73581f81e8fb6726f8b5102e412

`yarn testheadless test/production/middleware-environment-variables-in-node-server-reflect-the-usage-inference/index.test.ts` 
- middleware environment variables in node server reflect the usage inference > limits process.env to only contain env vars that are inferred from usage

<details>
<summary>Expand output</summary>

  ● middleware environment variables in node server reflect the usage inference › limits process.env to only contain env vars that are inferred from usage

    expect(received).toEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 1

      Object {
        "canBeInferred": "can-be-inferred",
        "rest": Object {
          "CAN_BE_INFERRED": "can-be-inferred",
          "X_CUSTOM_HEADER": "x-custom-header",
    +     "__NEXT_RUNTIME": "edge",
        },
      }

      40 |       parsed = JSON.parse(html)
      41 |     }).not.toThrow()
    > 42 |     expect(parsed).toEqual({
         |                    ^
      43 |       canBeInferred: 'can-be-inferred',
      44 |       rest: {
      45 |         CAN_BE_INFERRED: 'can-be-inferred',

      at Object.<anonymous> (production/middleware-environment-variables-in-node-server-reflect-the-usage-inference/index.test.ts:42:20)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| buildDuration | 15.3s | 15.2s | -148ms |
| buildDurationCached | 6.1s | 6s | -79ms |
| nodeModulesSize | 481 MB | 481 MB | ⚠️ +1.23 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.977 | 2.994 | ⚠️ +0.02 |
| / avg req/sec | 839.76 | 834.97 | ⚠️ -4.79 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.147 | 1.174 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 2179.86 | 2129.79 | ⚠️ -50.07 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.3 kB | 72.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.04 kB | 3.04 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.74 kB | 5.74 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16 kB | 16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| _buildManifest.js gzip | 461 B | 461 B | ✓ |
| Overall change | 461 B | 461 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| buildDuration | 17.3s | 17.4s | ⚠️ +136ms |
| buildDurationCached | 6.1s | 6.1s | -5ms |
| nodeModulesSize | 481 MB | 481 MB | ⚠️ +1.23 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.021 | 3.015 | -0.01 |
| / avg req/sec | 827.47 | 829.22 | +1.75 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.218 | 1.189 | -0.03 |
| /error-in-render avg req/sec | 2052.45 | 2102.08 | +49.63 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29.1 kB | 29.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.9 kB | 72.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.02 kB | 3.02 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.79 kB | 5.79 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.44 kB | 2.44 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 393 B | 393 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| _buildManifest.js gzip | 457 B | 457 B | ✓ |
| Overall change | 457 B | 457 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js process-edge | Change |
| - | - | - | - |
| index.html gzip | 528 B | 528 B | ✓ |
| link.html gzip | 542 B | 542 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.59 kB | 1.59 kB | ✓ |

</details>

</details>
Commit: 138d4e233701a73581f81e8fb6726f8b5102e412

File: packages/next/server/web/sandbox/context.ts [hunk 1]
```diff
@@ -266,7 +266,9 @@ function buildEnvironmentVariablesFrom(
   keys: string[]
 ): Record<string, string | undefined> {
   const pairs = keys.map((key) => [key, process.env[key]])
-  return Object.fromEntries(pairs)
+  const env = Object.fromEntries(pairs)
+  env.__NEXT_RUNTIME = 'edge'
```
Comment (ijjk) [inline packages/next/server/web/sandbox/context.ts:270]: Do we need to always expose this or can we only expose it if it's referenced like other environment variables in the edge runtime?

Comment (huozhi) [reply packages/next/server/web/sandbox/context.ts:270]: It's used inside the next server code to determine wether we use node or polyfill dependencies, so it's required at the very beginning that can't remove atm. I'll keep it in mind to see if we can opt it in the later versions

File: packages/next/build/webpack-config.ts [hunk 2]
```diff
@@ -2182,3 +2183,18 @@ export default async function getBaseWebpackConfig(
...
   return webpackConfig
 }
+
+function errorIfEnvConflicted(
+  config: NextConfigComplete,
+  key: string,
+  isServer: boolean
+) {
+  const isPrivateKey = /^(?:NODE_.+)|^(?:__.+)$/i.test(key)
+  const hasNextRuntimeKey = isServer && key === 'NEXT_RUNTIME'
```
Comment (ijjk) [inline packages/next/build/webpack-config.ts:2193]:
```suggestion
  const hasNextRuntimeKey = key === 'NEXT_RUNTIME'
```
Should we error if it's set for client build or server build for consistency?

**Review by ijjk — APPROVED** [review]

**Pull request merged**